Core/Instances: updated and enabled first common encounter frame packets

Ovahlord opened the pull request:
**Changes proposed**:
-  Updates and enables the basic packets for encounter frame for the new packet handling 
- implements handling for new encounter start and end opcodes
- implements battle resurrection charge pool for groups of any size

**Target branch(es)**: 6.x

**Known issues and TODO list**:
- Testing the battle rezz system for flaws and bugs
- add the remaining packets --> need sniffs for that if they even exist

Commit:
Core/Instances: updated and enabled first common encounter  frame packets

Commit:
fixed a typo

Shauren commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -164,6 +165,41 @@ namespace WorldPackets
+
+            WorldPacket const* Write() override;
+
+            Unit* unit = nullptr;
```
just stick ObjectGuid here (not even packed), not unit pointer

Commit:
Core/Instances: use ObjectGuid instead of unit pointers

DDuarte commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -164,6 +164,41 @@ namespace WorldPackets
+            WorldPacket const* Write() override;
+
+            ObjectGuid guid;
+            uint8 priority = 0; // used to update the position of the unit's current frame
```
The names should match what is in https://gist.github.com/DDuarte/2407219f73bfba69323a#file-structs_601_18179-h-L5686

So, `EncounterChangePriority`, not `SendEncounterChangePriority`; `Unit`, not `guid`; `TargetFramePriority`, not `priority`.

This applies to the other packets as well.

Shauren commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -164,6 +164,41 @@ namespace WorldPackets
+            ObjectGuid guid;
+            uint8 priority = 0; // used to update the position of the unit's current frame
+
+        };
```
class InstanceEncounterEngageUnit
    {
        ObjectGuid Unit;
        uint8 TargetFramePriority = 0;
    };

    class InstanceEncounterDisengageUnit
    {
        ObjectGuid Unit;
    };

    class InstanceEncounterChangePriority
    {
        ObjectGuid Unit;
        uint8 TargetFramePriority = 0;
    };

Please stick to these names

Commit:
Core/Instances: updated packet variable names to match packet standards

MitchesD commented on src/server/game/Instances/InstanceScript.cpp:
```diff
@@ -588,34 +589,51 @@ bool InstanceScript::CheckAchievementCriteriaMeet(uint32 criteria_id, Player con
     }
 
-    instance->SendToPlayers(&data);
+    // instance->SendToPlayers(&data);
```
it seems you don't need that line, if so, please remove it instead of commenting out

MitchesD commented on src/server/game/Server/Packets/InstancePackets.cpp:
```diff
@@ -117,3 +117,28 @@ WorldPacket const* WorldPackets::Instance::RaidInstanceMessage::Write()
+    return &_worldPacket;
+}
+
+
```
not needed empty line

Commit:
Core/Instances: get rid of client name part

MitchesD commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -164,6 +164,41 @@ namespace WorldPackets
+
+            ObjectGuid Unit;
+            uint8 TargetFramePriority = 0; // used to set the initial position of the frame if multiple frames are sent
+
```
again empty line

MitchesD commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -164,6 +164,41 @@ namespace WorldPackets
+
+            ObjectGuid Unit;
+            uint8 TargetFramePriority = 0; // used to update the position of the unit's current frame
+
```
once more

Commit:
cleanup

MitchesD commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -164,6 +164,41 @@ namespace WorldPackets
+            uint8 TargetFramePriority = 0; // used to update the position of the unit's current frame
+
+        };
+
```
and the last one

Commit:
correct some spacings

Shauren commented:
Will you be updating the remaining cases with the full packet dump or do you need sniffs for that? (not sure we can provide that, I don't remember if these packets are ever used)

Ovahlord commented:
I am currently using my iron docks heroic sniff which i've provided on the forums already and I don't have any other packet occurences there anymore. All I got left are the new  SMSG_INSTANCE_ENCOUNTER_START and END opcodes.

Commit:
Core/Instances: dropped semi depricated cases for SendEncounterUnit
*added functions to send encounter start and end packets to handle them seperate from the encounter unit since frames and encounters can start and end seperately

Commit:
fixed bad copy paste

Commit:
fixed build

Ovahlord commented:
Well I added the new  encounter opcodes with two new instance functions to handle them. Shall I also write a WoD encounter documentation to explain in which order the things are happening, including the scenario handling that is used for the instance quest?

Shauren commented:
I would rather have a generic way to handle that instance quest, not do it in scripts (obviously not in scope of this PR). 
As for using SendEncounterStart/End, that could be linked to BossAI/SetEncounterState (so preferably not in individual scrips again) so I think documenting it for script writers isn't neccessary

Ovahlord commented:
I was also thinking about putting this into the encounter states but that would probably need some script tweaks for multi-bosses so that they gonna use 1 boss state as common handler. I also need the information if those packets also get sent in earlier dungeons and raids such as TBC, WOTLK, Cata etc.

Commit:
Core/Instances: send encounter start and end packets uppon boss state changes. Values in sendEncounterStart are the same in all noticed packets so for now we gonna use magic numbers. Might be a great spot to add a db side encounter information table to set the battle rezz amount and recharge time

Commit:
Merge branch '6.x' of https://github.com/TrinityCore/TrinityCore into IronDocks

Lordron commented:
http://us.battle.net/wow/en/forum/topic/15690348911

something about in-combat-ressurection

Shauren commented on src/server/game/Instances/InstanceScript.cpp:
```diff
@@ -306,6 +307,19 @@ bool InstanceScript::SetBossState(uint32 id, EncounterState state)
+            switch (state)
+            {
+                case IN_PROGRESS:
+                    SendEncounterStart(1, 0, 1080000, 1080000); // Same values in all packets
```
The timer isn't constant - depends on number of players in raid (scales between 10 and 30), obviously constant for 5 man dungeons

Ovahlord replied:
Ye, just found the formular. While I am at it I could also add a battleRezz counter to the group class and use it to add the rezz restriction for the spell effects

Commit:
Core/Instances: added formular to calculate the battle rezz interval

Commit:
fixed build and added formular comment

Commit:
Core/Groups: add a group bound battle ressurection stacking counter to get battle rezz amount for future checks

Commit:
fioxed spelling mistakes

Commit:
fixed build

Commit:
Core/Spells: make battle ressurection spells only castable if the group has enough stacks available

Commit:
remove some spacings

Commit:
shuffled variables in group.h to pacify travis

tkrokli commented:
Just noticed that your commits do not link to your github account. If you want this feature to be available on github, set up your local git config with your github user e-mail address so your commits can link to your username + avatar properly (current commits don't have a user link).

Commit:
Core/Instances: move battle resurrection handling to instanceScript
todo: remove a battle rezz charge if ressurection has been accepted

Commit:
Core/Players: don't allow dead players to resurrect when there is no available resurrection charge

Commit:
fixed build and added brackets to prevent if else issues

Commit:
fixed remaining compile errors

Commit:
Core/Spells: treat battle resurrections as normal ones if the rezzing if the instance has no ongoing encounter right now

Commit:
Merge branch '6.x' of https://github.com/TrinityCore/TrinityCore into IronDocks

Commit:
Scripts/IronDocks: implemented basic abilities of Skulloc

Commit:
Scripts/IronDocks: more stuff for iron docks.

Shauren commented:
Merged, except the Iron Docks part - that should be in its own PR

Ovahlord commented:
sure, was accidentally pushed into the wrong branch. Gonna do a new PR if I make further progress
